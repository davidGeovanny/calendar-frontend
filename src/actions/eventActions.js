import Swal from "sweetalert2";
import { types } from "../types/types";
import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";

export const eventStartAddNew = ( event ) => {
    return async ( dispatch, getState ) => {

        const { uid, name } = getState().auth;

        try {
            const resp = await fetchWithToken( 'events', event, 'POST' );
            const body = await resp.json();
        
            if( body.ok ) {
                event.id = body.event.id;
                event.user = {
                    _id: uid,
                    name
                };

                dispatch( eventAddNew( event ) );
            }  
        } catch (error) {
            console.log(error);
        }
    };
};

const eventAddNew = ( event ) => {
    return {
        type: types.eventAddNew,
        payload: event
    };
}

export const eventSetActive = ( event ) => {
    return {
        type: types.eventSetActive,
        payload: event
    };
}

export const eventClearActiveEvent = () => {
    return {
        type: types.eventClearActiveEvent,
    };
}

export const eventStartUpdate = ( event ) => {
    return async ( dispatch ) => {
        try {
            const resp = await fetchWithToken( `events/${ event.id }`, event, 'PUT' );
            const body = await resp.json();

            if( body.ok ) {
                dispatch( eventUpdated( event ) );
            } else {
                Swal.fire({
                    title: 'Error',
                    text: body.msg,
                    icon: 'error'
                });
            }
        } catch (error) {
            console.log(error);
        }
    };
};

const eventUpdated = ( event ) => {
    return {
        type: types.eventUpdated,
        payload: event
    };
}

export const eventStartDelete = () => {
    return async ( dispatch, getState ) => {

        const { id } = getState().calendar.activeEvent;

        try {
            const resp = await fetchWithToken( `events/${ id }`, {}, 'DELETE' );
            const body = await resp.json();

            if( body.ok ) {
                dispatch( eventDeleted() );
            } else {
                Swal.fire({
                    title: 'Error',
                    text: body.msg,
                    icon: 'error'
                });
            }
        } catch (error) {
            console.log(error);
        }
    };
};

const eventDeleted = () => {
    return {
        type: types.eventDeleted,
    };
}

export const eventStartLoad = () => {
    return async ( dispatch ) => {
        try {
            const resp = await fetchWithToken( 'events' );
            const body = await resp.json();

            const events = prepareEvents( body.events );

            dispatch( eventLoaded( events ) );
        } catch (error) {
            console.log(error);
        }
    };
}

const eventLoaded = ( events ) => {
    return {
        type: types.eventLoaded,
        payload: events
    }
};

export const eventLogout = () => {
    return {
        type: types.eventLogout,
    }
};