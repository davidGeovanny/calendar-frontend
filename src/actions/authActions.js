import Swal from "sweetalert2";
import { types } from "../types/types";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { eventLogout } from "./eventActions";

export const startLogin = ( email, password ) => {
    return async ( dispatch ) => {
        const resp = await fetchWithoutToken( 'auth/', { email, password }, 'POST' );
        const body = await resp.json();

        if( body.ok ) {
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login(
                { 
                    uid: body.uid, 
                    name: body.name 
                }
            ));
        } else {
            Swal.fire({
                title: 'Error',
                text: body.msg,
                icon: 'error' 
            });
        }
    };
};

const login = ( user ) => {
    return {
        type: types.authLogin,
        payload: user
    };
};

export const startRegister = ( name, email, password ) => {
    return async ( dispatch ) => {

        const resp = await fetchWithoutToken( 'auth/new', { name, email, password }, 'POST' );

        const body = await resp.json();

        if( body.ok ) {
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login(
                { 
                    uid: body.uid, 
                    name: body.name 
                }
            ));
        } else {
            Swal.fire({
                title: 'Error',
                text: body.msg,
                icon: 'error',
            });
            dispatch( checkingFinish() );
        }
    };
};

export const startChecking = () => {
    return async ( dispatch ) => {
        const resp = await fetchWithToken( 'auth/renew' );

        const body = await resp.json();

        if( body.ok ) {
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login(
                { 
                    uid: body.uid, 
                    name: body.name 
                }
            ));
        } else {
            dispatch( checkingFinish() );
        }
    };
};

const checkingFinish = () => {
    return {
        type: types.authCheckingFinish,
    };
};

export const startLogout = () => {
    return async ( dispatch ) => {
        localStorage.clear();
        dispatch( logout() );
        dispatch( eventLogout() );
    };
};

const logout = () => {
    return {
        type: types.authLogout,
    };
};