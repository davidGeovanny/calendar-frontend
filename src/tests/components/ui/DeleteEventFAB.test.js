import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { DeleteEventFAB } from '../../../components/ui/DeleteEventFAB';
import { eventStartDelete } from '../../../actions/eventActions';

jest.mock( '../../../actions/eventActions', () => {
    return {
        eventStartDelete: jest.fn(),
    };
});

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const initState = {};

const store = mockStore( initState );
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <DeleteEventFAB />
    </Provider>
);

describe('Test with DeleteEventFAB component', () => {
    
    test('should match with the snapshot', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should call the event eventStartDelete when the button is clicked', () => {
        wrapper.find('button').simulate('click', {
            preventDefault() {}
        });

        expect( eventStartDelete ).toHaveBeenCalled();
    });
    
    
});
