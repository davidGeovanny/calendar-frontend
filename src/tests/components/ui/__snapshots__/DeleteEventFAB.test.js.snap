// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test with DeleteEventFAB component should match with the snapshot 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
  <span>
     Eliminar 
  </span>
</button>
`;
